underfs/s3a/src/main/java/alluxio/underfs/s3a/S3AInputStream.java

```diff
@@ -0,0 +1,137 @@
+import java.io.InputStream;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * A wrapper around an {@link S3ObjectInputStream} which handles skips efficiently.
+ */
+@NotThreadSafe
+public class S3AInputStream extends InputStream {
+  /** Client for operations with s3. */
+  private final AmazonS3 mClient;
+  /** Name of the bucket the object resides in. */
+  private final String mBucketName;
+  /** The path of the object to read. */
+  private final String mKey;
+
+  /** The backing input stream from s3. */
+  private S3ObjectInputStream mIn;
+  /** The current position of the stream. */
+  private long mPos;
+
+  /**
+   * Constructor for a input stream of an object in s3 using the aws-sdk implementation to read
+   * the data. The stream will be positioned at the start of the file.
+   *
+   * @param bucketName the bucket the object resides in
+   * @param key the path of the object to read
+   * @param client the s3 client to use for operations
+   */
+  public S3AInputStream(String bucketName, String key, AmazonS3 client) {
+    this(bucketName, key, client, 0L);
+  }
+
+  /**
+   * Constructor for a input stream of an object in s3 using the aws-sdk implementation to read
+   * the data. The stream will be positioned at the specified position.
+   *
+   * @param bucketName the bucket the object resides in
+   * @param key the path of the object to read
+   * @param client the s3 client to use for operations
+   * @param position the position to begin reading from
+   */
+  public S3AInputStream(String bucketName, String key, AmazonS3 client, long position) {
+    mBucketName = bucketName;
+    mKey = key;
+    mClient = client;
+    mPos = position;
+  }
+
+  @Override
+  public void close() {
+    closeStream();
+  }
+
+  @Override
+  public int read() throws IOException {
+    if (mIn == null) {
+      openStream();
+    }
+    int value = mIn.read();
+    if (value != -1) { // valid data read
+      mPos++;
+    }
+    return value;
+  }
+
+  @Override
+  public int read(byte[] b) throws IOException {
+    return read(b, 0, b.length);
+  }
+
+  @Override
+  public int read(byte[] b, int offset, int length) throws IOException {
+    if (length == 0) {
+      return 0;
+    }
+    if (mIn == null) {
+      openStream();
+    }
+    int read = mIn.read(b, offset, length);
+    if (read != -1) {
+      mPos += read;
+    }
+    return read;
+  }
+
+  @Override
+  public long skip(long n) {
+    if (n <= 0) {
+      return 0;
+    }
+    closeStream();
+    mPos += n;
+    openStream();
+    return n;
+  }
+
+  /**
+   * Opens a new stream at mPos if the wrapped stream mIn is null.
+   */
+  private void openStream() {
+    if (mIn != null) { // stream is already open
+      return;
+    }
+    GetObjectRequest getReq = new GetObjectRequest(mBucketName, mKey);
+    getReq.setRange(mPos);
+    mIn = mClient.getObject(getReq).getObjectContent();
+  }
+
+  /**
+   * Closes the current stream.
+   */
+  private void closeStream() {
```

Comment:
Abort prevents the HTTP library from reusing the connection. I left a TODO here to optimize it, but I'm not sure if it is necessary as I have not benchmarked the performance gain by reusing http connections.